pkg/util/cluster/cluster.go

```diff
@@ -345,22 +352,111 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
 	return nil
 }
 
-func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, workerSubnet string) {
-	// pick a random 23 in range [10.3.0.0, 10.127.255.0]
+// ipRangesContainCIDR checks, weather any of the ipRanges overlap with the cidr string. In case cidr isn't valid, false is returned.
+func ipRangesContainCIDR(ipRanges []*net.IPNet, cidr string) (bool, error) {
+	_, cidrNet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return false, err
+	}
+
+	for _, snet := range ipRanges {
+		if snet.Contains(cidrNet.IP) || cidrNet.Contains(snet.IP) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// GetIPRangesFromSubnet converts a given azure subnet to a list if IPNets.
+// Because an az subnet can cover multiple ipranges, we need to return a slice
+// instead of just a single ip range. This function never errors. If something
+// goes wrong, it instead returns an empty list.
+func GetIPRangesFromSubnet(subnet mgmtnetwork.Subnet) []*net.IPNet {
+	ipRanges := []*net.IPNet{}
+	if subnet.AddressPrefix != nil {
+		_, ipRange, err := net.ParseCIDR(*subnet.AddressPrefix)
+		if err == nil {
+			ipRanges = append(ipRanges, ipRange)
+		}
+	}
+
+	if subnet.AddressPrefixes == nil {
+		return ipRanges
+	}
+
+	for _, snetPrefix := range *subnet.AddressPrefixes {
+		_, ipRange, err := net.ParseCIDR(snetPrefix)
+		if err == nil {
+			ipRanges = append(ipRanges, ipRange)
+		}
+	}
+	return ipRanges
+}
+
+// getAllDevSubnets queries azure to retrieve all subnets assigned the vnet
+// `dev-vnet` in the current resource group
+func (c *Cluster) getAllDevSubnets() ([]mgmtnetwork.Subnet, error) {
+	allSubnets := []mgmtnetwork.Subnet{}
+	availSnetResults, err := c.subnets.List(context.Background(), c.env.ResourceGroup(), "dev-vnet")
+	if err != nil {
+		return allSubnets, err
+	}
+	allSubnets = append(allSubnets, availSnetResults.Values()...)
+	for availSnetResults.NotDone() {
+		err = availSnetResults.NextWithContext(context.Background())
+		if err != nil {
+			break
+		}
+		allSubnets = append(allSubnets, availSnetResults.Values()...)
+	}
+
+	return allSubnets, nil
+}
+
+func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, workerSubnet string, err error) {
+	// pick a random 23 in range [10.3.0.0, 10.127.255.0], making sure it doesn't
```

Comment:
idea: I wonder if this randomness is actually what shoots us in the foot. What if instead, we got very structured with this.

What if, given that both masters and workers are /24 nets, we adopted a CIDR convention rather than shooting in the dark x many times? I'm thinking something like the 10.3.0.0/ space is entirely dedicated to dev clusters, and we take that a slice at a time:

```
10.3.0.0/24
10.3.1.0/24
10.3.2.0/24
10.3.3.0/24
10.3.4.0/24
10.3.5.0/24
10.3.6.0/24
...
```

Now, to find the next "open" subnet, all you have to do is iterate over this CIDR convention, and see if it shows up in the VNet or not, and it supports a scale I think we can easily handle (255 subnets, or 127 clusters per region). Thoughts?

Reply:
It was discussed here as a potential follow-up improvement: https://github.com/Azure/ARO-RP/pull/3706#discussion_r1684208163